suprb/optimizer/rule/ns/novelty_calculation.py

```diff
@@ -0,0 +1,100 @@
+from abc import ABCMeta, abstractmethod
+
+import numpy as np
+from suprb.rule import Rule
+from suprb.base import BaseComponent
+from scipy.spatial.distance import hamming
+from .novelty_search_type import NoveltySearchType
+from .archive import Archive
+
+
+class NoveltyCalculation(BaseComponent, metaclass=ABCMeta):
+    def __init__(self, novelty_search_type: NoveltySearchType, archive: Archive, k_neighbor: int):
+        self.novelty_search_type = novelty_search_type
+        self.archive = archive
+        self.k_neighbor = k_neighbor
+
+    def __call__(self, rules: list[Rule]) -> list[Rule]:
+        return self._novelty_score(rules=rules)
+
+    def _add_distances_to_archive_rules(self, rule: Rule, archive: list[Rule]):
+        for i, _ in enumerate(archive):
+            archive[i].distance = hamming(rule.match_, archive[i].match_)
+
+    def _novelty_score(self, rules: list[Rule]) -> list[Rule]:
+        """ The basic novely calculation based on the hamming distance. 
+            Takes every rule and calculates the hamming distance to each rule contained in the archive
+            and averages the distances of the k-nearest neighbors to get the novelty score.
+        """
+        novelty_search_rules = []
+        archive = self.archive.get_archive()[:]
+        archive.extend(rules)
+
+        for rule in self.novelty_search_type.filter_rules(rules):
+            self._add_distances_to_archive_rules(rule, archive)
+
+            archive = sorted(archive, key=lambda archive_rule: archive_rule.distance)
+            distances = [sorted_rule.distance for sorted_rule in archive]
+
+            local_competition = self.novelty_search_type.local_competition(rule, archive)
+            novelty_score = local_competition + self._novelty_score_calculation(distances, rule.fitness_)
+
+            rule.novelty_score = novelty_score
+            novelty_search_rules.append(rule)
+
+        return novelty_search_rules
+
+    def _novelty_score_calculation(self, distances: list[float], fitness: float):
```

Comment:
fitness is unused

Reply:
Yes that is kind of unfortunate, but necessary because of the `NoveltyFitnessBiased` calculation. Here we need the rules fitness for `_novelty_score_calculation`. Since the same `_novelty_score` logic is used in all subclasses (to avoid code duplication), I chose to add the fitness argument here, although it is only used for the `NoveltyFitnessBiased`

Reply:
I see. So for method overloading we need to have the same signature as well as name in python?

Reply:
Yes

Reply:
Can't we use **kwargs, though?

Reply:
True. I change it